halt.8
.\"{{{}}}
.\"{{{  Title
.TH HALT,REBOOT 8 "May 13, 1993" "" "Linux Programmer's Manual"
.\"}}}
.\"{{{  Name
.SH NAME
halt, reboot \- stop the system.
.\"}}}
.\"{{{  Synopsis
.SH SYNOPSIS
.B /etc/halt
.RB [ \-t
.IR sec ]
.RB [ \-nq ]
.br
.B /etc/reboot
.RB [ \-t
.IR sec ]
.RB [ \-nq ]
.\"}}}
.\"{{{  Description
.SH DESCRIPTION
\fBHalt\fP and \fBreboot\fP send all processes the warning signal, then
wait for 10 seconds, then send all processes the kill signal.  After
that the system is either halted or rebooted.
.\"}}}
.\"{{{  Options
.SH OPTIONS
.IP \fB\-n\fP
Don't sync before reboot or halt.
.IP \fB\-q\fP
Do a `quick' reboot or halt, don't send the warning signal.
.IP "\fB\-t\fP \fIsec\fP"
Wait \fIsec\fP seconds between sending processes the warning and the
kill signal.
.\"}}}
.\"{{{  Diagnostics
.SH DIAGNOSTICS
If you're not the superuser, you will get the message `Can't idle init'.
.\"}}}
.\"{{{  Notes
.SH NOTES
\fIReboot\fP and \fIhalt\fP shouldn't be called directly, but only by a
frontend like \fIshutdown\fP(1).  Even better, they should be builtin to
\fIshutdown\fP.  However, historical practice prevents this.
.PP
\fIHalt\fP and \fIreboot\fP are just links to the same program, so the
man page should be installed twice, as \fBhalt.8\fP and as
\fBreboot.8\fP.
.\"}}}
.\"{{{  Author
.SH AUTHOR
Miquel van Smoorenburg, miquels@drinkel.nl.mugnet.org
.\"}}}
.\"{{{  See also
.SH "SEE ALSO"
shutdown(8), init(1)
.\"}}}


History
 Version:	1.0, 01-Feb-92
		- Initial version, very primitive for the Minix
		  operating system. Required some mods. to the
		  kernel.

 		1.1, 30-Apr-92
		- Read manual wrong: there is no 'action' field called
		  process, but all entries are of type process. Every
		  'process' get exec'ed by /bin/sh -c 'exec command'.
		- Rapidly respawning processes are caught in the act.
		- _SYSV support is really Linux support,
		  done by poe@daimi.aau.dk on 25-Mar-92.

		1.2, 16-Jun-92
		- Bugreport from Michael Haardt ; removed deadlock
		  and added 'waitpid' instead of 'wait' for SYSV.

		1.3, 05-Jul-92
		- Got a 386, so installed Linux. Added 'soft' reboot
		  to be default under linux. Fixed some typos.

		2.0 08-Dec-92
		- Rewrote the code totally, so started with a new
		  version number.
		- Dropped Minix support, this code now is Linux - specific.
		- With TEST switch on, this init & telinit can
		  run standalone for test purposes.

		2.1 28-Jan-93
		- Fixed a bug with 'boot' and 'once'.
		- Check 'initdefault' for validity.
		- Reckognizes "single" as command line argument.
		- Retries execvp with 'sh -c exec ..' if command
		  is a shell script. (shouldn't execvp do this?)
		- Added patches to use syslog if defined.

		2.2 02-Mar-93
		- Made wait()'s asynchronous
		- Changed whole thing to one big state machine
		- Now using 'pseudo levels' # & * for SYSINIT & BOOT
		- Added a new type of 'action', called leave. This
		  process will be executed when the system goes from a
		  runlevel specified in it's runlevel field to a 
		  level that's not. Nice to bring down NFS and the like.

		2.3 24-Mar-93
		- Ripped out the 'leave' action. To difficult, and
		  unneeded.
		- Going single user now kills _all_ processes.
		- Added '-t secs' option to all commands.
		- This version is stable enough to post.

		2.4a 22-Apr-93
		- Fixed last to reckognize BSD style wtmp logging.
		- Changed init to write wtmp records that are
		  SysV compliant but are also reckognized by the
		  BSD last. Added a '+' option to the 'process'
		  field of inittab, for getties that want to do
		  their own utmp/wtmp housekeeping (kludge!).
		- Now accepts a runlevel on the command line,
		  and reckognizes the 'auto' argument. (Sets the
		  environment variable AUTOBOOT to YES)

		2.4g 15-May-93
		- Changed init to really catch SIGPWR 'cause we
		  hooked up an UPS to the Linux machine. The
		  keyword for catching the TreeFingerSalute is
		  now "ctrlaltdel" instead of "power{wait,fail}".

		2.4  24-May-93
		- Send out the official version into the world as
		  SysVinit-2.4.tar.z.


halt.c
/*
 * Halt		Stop the system running.
 *		Halt stops init, and then kills all processes.
 *		After that, the program (shell script) /etc/rc.d/rc.0
 *		is executed. If that can't be found, /etc/brc is used.
 *
 * Usage:	halt [-n][-q][-t secs]
 *		-n: don't sync before halting the system
 *		-q: quick reboot, don't send processes a warning signal.
 *		-t secs: delay between SIGTERM and SIGKILL
 *
 * Reboot	Identical to halt, but after rc.0 has been
 *		executed, the system is automatically rebooted.
 *
 *		Reboot and halt are both this program. Reboot
 *		is just a link to halt.
 *
 * Author:	Miquel van Smoorenburg, miquels@drinkel.nl.mugnet.org
 *
 * Version:	1.3,  24-03-1993
 *
 */

#include <sys/types.h>
#include <stdlib.h>
#include <signal.h>
#include <utmp.h>
#include <fcntl.h>
#include <string.h>
#include <unistd.h>
#include <sys/wait.h>
#include <time.h>
#include <stdio.h>

char *Version = "@(#) halt 1.3 24-03-1993 MvS";

/* To keep this thing less system dependant, check some things */
#ifndef UTMP
#  ifdef UTMP_FILE
#    define UTMP UTMP_FILE		/* The real name */
#    define WTMP WTMP_FILE
#  else
#    define UTMP "/etc/utmp"
#    define WTMP "/usr/adm/wtmp"
#  endif
#endif
#ifndef NO_PROCESS
#  define NO_PROCESS 0
#endif
#ifndef _NSIG
#  define _NSIG NSIG
#endif

#define INITPID		1	/* PID of init				*/
#define DOWN1 "/etc/rc.d/rc.0"	/* Script to be run if system is down	*/
#define DOWN2 "/etc/brc"	/* Second choise */
#define NO_KILL_ALL	0	/* Linux can kill all processes simult. */
extern int getopt();		/* For getopt (3) 			*/
extern int optind;		/* Getopt too				*/
int nosync = 0;			/* Don't sync before reboot or halt	*/
int quick = 0;			/* Quick kill of all processes		*/
char *progname;			/* Invoked as 'halt' or 'reboot' ?	*/
int sltime = 20;		/* Sleeptime between TERM and KILL	*/

/*
 * Sleep 't' seconds. Print a dot every second.
 */
void dotdot(t)
int t;
{
  int f;
  
  for(f = 0; f < t; f++) {
  	sleep(1);
  	putchar('.');
  	fflush(stdout);
  }
  printf("\r\n");
}

#if NO_KILL_ALL
/*
 * Kill all processes spawned by init by reading /etc/utmp.
 */
void killall()
{
  struct utmp utmp;
  int fd;
  int pid = getpid();

  if ((fd = open(UTMP, O_RDONLY)) < 0) {
  	fprintf(stderr, "halt: cannot open %s\n", UTMP);
  	exit(1);
  }
  if (!quick) {
	/* Send all processes the SIGTERM signal */
	printf("\r%s: sending all processes the TERM signal", progname);
	while(read(fd, &utmp, sizeof(utmp)) == sizeof(utmp)) {
  		if (utmp.ut_type && utmp.ut_pid && utmp.ut_pid != pid)
  			kill(utmp.ut_pid, SIGTERM);
	}
	dotdot(sltime); /* Give processes the time to clean up */
	lseek(fd, 0L, SEEK_SET);
  }

  printf("%s: sending all processes the KILL signal..\r\n", progname);
  /* Send all processes the SIGKILL signal */
  while(read(fd, &utmp, sizeof(utmp)) == sizeof(utmp)) {
  	if (utmp.ut_type && utmp.ut_pid && utmp.ut_pid != pid)
  		kill(utmp.ut_pid, SIGKILL);
  }
  close(fd);
}
#else
/*
 * Kill all processes with the special argument '-1' to kill().
 */
void killall()
{
  if (!quick) {
	printf("\r%s: sending all processes the TERM signal", progname);
	kill(-1, SIGTERM);
	dotdot(sltime);
  }	
  printf("%s: sending all processes the KILL signal..\r\n", progname);
  kill(-1, SIGKILL);
}
#endif

/*
 * Send usage message.
 */
void usage()
{
  fprintf(stderr, "usage: %s [-qn]\n", progname);
  exit(1);
}

/*
 * Main program.
 * Kill all processes, write a wtmp entry and reboot cq. halt.
 */
int main(argc, argv)
int argc;
char **argv;
{
  struct utmp wtmp;
  int c, pid, st, fd, f;
  time_t t;
  int do_reboot = 0;

  /* Ignore all signals */
  for(f = 1; f < _NSIG; f++) {
#ifdef SIGCHLD
  if (f == SIGCHLD) continue;
#endif
#ifdef SIGCLD
  if (f == SIGCLD)  continue;
#endif
	signal(f, SIG_IGN);
  }

  /* Find out who we are */
  if ((progname = strrchr(argv[0], '/')) != NULL)
  	progname++;
  else
  	progname = argv[0];
  	
  if (!strcmp(progname, "reboot")) do_reboot = 1;

  /* Get flags */
  while((c = getopt(argc, argv, "nqt:")) != EOF) {
  	switch(c) {
  		case 'n':
  			nosync = 1;
  			break;
  		case 'q':
  			quick = 1;
  			break;
		case 't':
			sltime = atoi(optarg);
			break;
  		default:
  			usage();
  	}
  }
  if (optind < argc) usage();
  
  /* Tell init to stop */
  if (kill(INITPID, SIGTSTP) != 0) {
  	fprintf(stderr, "%s: can't idle init\n", progname);
  	exit(1);
  }

  /* Kill all processes */
  killall();
  
  if (!nosync) sync();

  /* Record the fact that's we're going down */
  if ((fd = open(WTMP, O_WRONLY|O_APPEND)) >= 0) {
  	time(&t);
  	strcpy(wtmp.ut_user, "shutdown");
  	strcpy(wtmp.ut_line, "~");
  	strcpy(wtmp.ut_id,  "~~");
  	wtmp.ut_pid = 0;
  	wtmp.ut_type = RUN_LVL;
  	wtmp.ut_time = t;
  	write(fd, (char *)&wtmp, sizeof(wtmp));
  	close(fd);
  }

  /* Now we're alone, execute /etc/brc */
  if ((pid = fork()) == 0) {
	execlp(DOWN1, DOWN1, (char *)NULL);
	execlp(DOWN2, DOWN2, (char *)NULL);
	exit(1);
  }
  if (pid > 0)
	wait(&st);
  else
	fprintf(stderr, "Warning: cannot fork: %s not executed\n",
	    	DOWN1);

  if (!nosync) sync();
  if (do_reboot) {
	if (!nosync) sleep(2); /* For sync */
	reboot(0xfee1dead, 672274793, 0x01234567);
  } else
	printf("\r\nThe system is halted\r\n");
  exit(0);
}
